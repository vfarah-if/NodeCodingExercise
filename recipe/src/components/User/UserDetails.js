import React, { useEffect, useState } from 'react';
import PropTypes from 'prop-types';
import './user.css';

// Utilising https://reqres.in/api/users/{id} to get fake real user details
export default function UserDetails({ id }) {
	const [user, setUser] = useState(null);
	const [isLoaded, setIsLoaded] = useState(false);
	const fetchUserData = async (id) => {
		const response = await fetch(`https://reqres.in/api/users/${id}`);
		const { data } = await response.json();
		console.debug(response, data);
		setUser(data);
	};

	useEffect(() => {
		setIsLoaded(true);
		return fetchUserData(id);
	}, [id]);

	if (!user && !isLoaded) {
		return 'loading ...';
	}

	if (!user && isLoaded) {
		return (
			<details>
				<p>No user details for User id '{id}'</p>
			</details>
		);
	}

	return (
		<details>
			<summary>
				{user?.first_name} {user?.last_name} @ {user?.email}
			</summary>
			<img src={user?.avatar} alt="Avatar"></img>
			<p>User BIO for User id '{id}'</p>
		</details>
	);
}

UserDetails.propTypes = {
	id: PropTypes.number.isRequired,
};
