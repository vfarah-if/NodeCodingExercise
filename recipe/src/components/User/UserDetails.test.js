import React from 'react';
import { render, unmountComponentAtNode } from 'react-dom';
import { act } from 'react-dom/test-utils';

import UserDetails from './UserDetails';
import UserDataJsonResponse from './UserDetails.test.json';

describe('UserDetails', () => {
	let container = null;
	beforeEach(() => {
		container = document.createElement('div');
		document.body.appendChild(container);
	});

	afterEach(() => {
		unmountComponentAtNode(container);
		container.remove();
		container = null;
	});

	test('should have fake data to start with', () => {
		expect(UserDataJsonResponse).toBeDefined();
		expect(UserDataJsonResponse.data).toBeDefined();
	});

	test('should get user details when valid id is supplied', async () => {
		const { id } = UserDataJsonResponse.data;
		console.debug(id);
		jest.spyOn(global, 'fetch').mockImplementation(() =>
			Promise.resolve({
				json: () => Promise.resolve(UserDataJsonResponse),
			})
		);

		await act(async () => {
			render(<UserDetails id={id}></UserDetails>, container);
		});

		expect(container.querySelector('details')).toMatchSnapshot();
	});

	test('should get no user found when user response is empty', async () => {
		jest.spyOn(global, 'fetch').mockImplementation(() =>
			Promise.resolve({
				json: () => Promise.resolve({}),
			})
		);

		await act(async () => {
			render(<UserDetails id="1000"></UserDetails>, container);
		});

		expect(container.querySelector('details')).toMatchSnapshot();
	});
});
