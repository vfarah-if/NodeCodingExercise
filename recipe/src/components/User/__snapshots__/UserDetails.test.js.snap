// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserDetails should get no user found when user response is empty 1`] = `
<details>
  <p>
    No user details for User id '
    1000
    '
  </p>
</details>
`;

exports[`UserDetails should get user details when valid id is supplied 1`] = `
<details>
  <summary>
    Janet
     
    Weaver
     @ 
    janet.weaver@reqres.in
  </summary>
  <img
    alt="Avatar"
    src="https://reqres.in/img/faces/2-image.jpg"
  />
  <p>
    User BIO for User id '
    2
    '
  </p>
</details>
`;
